### DISCUSSION TITLE
Mini.Extra `gen_ai_spec.diagnostic` with CoC

### DISCUSSION THREAD

**Discussion by wroyca** [nvim-mini/mini.nvim#992] [Show and tell]
Here's Mini.Extra `gen_ai_spec.diagnostic = function()` adapted to work with CoC:

```lua
local gen_ai_spec = require [[mini.extra]].gen_ai_spec

--- Current buffer diagnostic textobject
---
--- Notes:
--- - Both `a` and `i` textobjects return |vim.fn["CocAction"]("diagnosticList")| output for the
---   current buffer. It is modified to fit |MiniAi-textobject-specification|.
---
---@return __extra_ai_spec_return
gen_ai_spec.diagnostic = function()
  return function(ai_type)
    local cur_diag = vim.fn["CocAction"]("diagnosticList")
    local regions = {}
    for _, diag in ipairs(cur_diag) do
      local from = { line = diag.lnum, col = diag.col }
      local to = { line = diag.end_lnum, col = diag.end_col - 1 }
      if to.line == nil or to.col == nil then to = { line = diag.lnum , col = diag.col - 1 } end
      table.insert(regions, { from = from, to = to })
    end
    return regions
  end
end
```

**Comment by echasnovski** [comment]
Thanks for sharing!

I used to use 'coc.nvim' for a brief time, but switched to built-in diagnostics. Which seems also to be a better overall approach for 'mini.nvim'.

Also, there doesn't seem to be a reason to override `gen_ai_spec.diagnostic` for this to be used in the 'mini.ai' spec. A less nested `coc_diagnostic_spec = function(ai_type) ... end` should be enough.

**Reply by wroyca** [reply]
> I used to use 'coc.nvim' for a brief time, but switched to built-in diagnostics. Which seems also to be a better overall approach for 'mini.nvim'.

I don't really have a strong opinion on this; both options are excellent. :)

Personally, I've grown to trust @fannheyward and the community for their excellent work on CoC, which addresses a much broader range of needs beyond *"just"* LSP alone. 

> Also, there doesn't seem to be a reason to override gen_ai_spec.diagnostic for this to be used in the 'mini.ai' spec. A less nested coc_diagnostic_spec = function(ai_type) ... end should be enough.

Correct! 

I chose to do it that way in my own setup to maintain coherence in my configuration. My approach isn't a strict requirement, of course. 🍻

**Comment by fannheyward** [comment]
You can now use `diagnostic.displayByVimDiagnostic`, enable this configuration will set diagnostics to nvim's `vim.diagnostic` that mini.nvim can request to use in Pick or statusline etc.
